File: android/sandwich/src/main/kotlin/io/qonversion/sandwich/NoCodesSandwich.kt
package io.qonversion.sandwich

import android.content.Context
import android.util.Log
import io.qonversion.nocodes.NoCodes
import io.qonversion.nocodes.NoCodesConfig
import io.qonversion.nocodes.dto.LogLevel
import io.qonversion.nocodes.interfaces.NoCodesDelegate
import io.qonversion.nocodes.interfaces.ScreenCustomizationDelegate
import io.qonversion.nocodes.dto.QScreenPresentationConfig
import io.qonversion.nocodes.dto.QAction
import io.qonversion.nocodes.error.NoCodesError
import androidx.core.content.edit

class NoCodesSandwich {

    private var defaultPresentationConfig: QScreenPresentationConfig? = null
    private val screenPresentationConfigs = mutableMapOf<String, QScreenPresentationConfig>()
    private var isCustomizationDelegateSet = false
    private val screenCustomizationDelegate = object : ScreenCustomizationDelegate {
        override fun getPresentationConfigurationForScreen(contextKey: String): QScreenPresentationConfig {
            return screenPresentationConfigs[contextKey] ?: defaultPresentationConfig ?: QScreenPresentationConfig()
        }
    }
    private lateinit var noCodesDelegate: NoCodesDelegate

    // region Initialization

    fun initialize(
        context: Context,
        projectKey: String,
        proxyUrl: String? = null,
        logLevelKey: String? = null,
        logTag: String? = null
    ) {
        val configBuilder = NoCodesConfig.Builder(context, projectKey)

        proxyUrl?.let {
            configBuilder.setProxyURL(it)
        }

        logLevelKey?.let {
            try {
                val logLevel = LogLevel.valueOf(it)
                configBuilder.setLogLevel(logLevel)
            } catch (e: IllegalArgumentException) {
                Log.w("No-Codes Sandwich", "Invalid log level provided: $it")
            }
        }

        logTag?.let {
            configBuilder.setLogTag(it)
        }

        NoCodes.initialize(configBuilder.build())
    }

    fun storeSdkInfo(context: Context, source: String, version: String) {
        context.getSharedPreferences(NO_CODES_PREFS_NAME, Context.MODE_PRIVATE).edit {
            putString(NO_CODES_KEY_VERSION, version)
            putString(NO_CODES_KEY_SOURCE, source)
        }
    }

    // endregion

    // region Delegate

    fun setDelegate(eventListener: NoCodesEventListener) {
        noCodesDelegate = createNoCodesDelegate(eventListener)
        NoCodes.shared.setDelegate(noCodesDelegate)
    }

    fun setScreenCustomizationDelegate() {
        if (!isCustomizationDelegateSet) {
            isCustomizationDelegateSet = true
            NoCodes.shared.setScreenCustomizationDelegate(screenCustomizationDelegate)
        }
    }

    fun setScreenPresentationConfig(configData: Map<String, Any?>, contextKey: String? = null) {
        val config = configData.toScreenPresentationConfig()

        if (!isCustomizationDelegateSet) {
            isCustomizationDelegateSet = true
            NoCodes.shared.setScreenCustomizationDelegate(screenCustomizationDelegate)
        }

        contextKey?.let {
            screenPresentationConfigs[contextKey] = config
        } ?: run {
            screenPresentationConfigs.clear()
            defaultPresentationConfig = config
        }
    }

    // endregion

    // region Screen Management

    fun showScreen(contextKey: String) {
        NoCodes.shared.showScreen(contextKey)
    }

    fun close() {
        NoCodes.shared.close()
    }

    // endregion

    // region Configuration

    fun setLogLevel(logLevelKey: String) {
        val logLevel = LogLevel.valueOf(logLevelKey)
        NoCodes.shared.setLogLevel(logLevel)
    }

    fun setLogTag(logTag: String) {
        NoCodes.shared.setLogTag(logTag)
    }

    // endregion

    // region Private

    private fun createNoCodesDelegate(eventListener: NoCodesEventListener): NoCodesDelegate {
        return object : NoCodesDelegate {
            override fun onScreenShown(screenId: String) {
                val payload = mapOf("screenId" to screenId)
                eventListener.onNoCodesEvent(NoCodesEventListener.Event.ScreenShown, payload)
            }

            override fun onActionStartedExecuting(action: QAction) {
                eventListener.onNoCodesEvent(NoCodesEventListener.Event.ActionStarted, action.toMap())
            }

            override fun onActionFailedToExecute(action: QAction) {
                eventListener.onNoCodesEvent(NoCodesEventListener.Event.ActionFailed, action.toMap())
            }

            override fun onActionFinishedExecuting(action: QAction) {
                eventListener.onNoCodesEvent(NoCodesEventListener.Event.ActionFinished, action.toMap())
            }

            override fun onFinished() {
                eventListener.onNoCodesEvent(NoCodesEventListener.Event.Finished)
            }

            override fun onScreenFailedToLoad(error: NoCodesError) {
                eventListener.onNoCodesEvent(NoCodesEventListener.Event.ScreenFailedToLoad, error.toMap())
            }
        }
    }

    // endregion
}

File: ios/sandwich/NoCodesSandwich.swift
//
//  NoCodesSandwich.swift
//  QonversionSandwich
//
//  Created by Suren Sarkisyan on 08.05.2025.
//  Copyright © 2025 Qonversion Inc. All rights reserved.
//
#if os(iOS)

import Foundation
import Qonversion
import UIKit

public class NoCodesSandwich: NSObject {
    private var noCodesEventListener: NoCodesEventListener?
    private var defaultPresentationConfig: NoCodesPresentationConfiguration? = nil
    private var screenPresentationConfigs: [String: NoCodesPresentationConfiguration] = [:]
    private var isCustomizationDelegateSet = false
    
    @objc public init(noCodesEventListener: NoCodesEventListener) {
        self.noCodesEventListener = noCodesEventListener
    }

    @objc public func initialize(projectKey: String) {
        let noCodesConfig = NoCodesConfiguration(projectKey: projectKey)
        
        NoCodes.initialize(with: noCodesConfig)
        NoCodes.shared.set(delegate: self)
    }
  
    @objc public func storeSdkInfo(source: String, version: String) {
        // Does nothing on iOS as No-Codes are integrated into the Qonversion SDK and share its source and version
    }
    
    @MainActor @objc public func setScreenPresentationConfig(_ configData: [String: Any], forContextKey contextKey: String? = nil) {
        let config = configData.toPresentationConfig()
        
        if (!isCustomizationDelegateSet) {
            isCustomizationDelegateSet = true
            NoCodes.shared.set(screenCustomizationDelegate: self)
        }
        
        if let contextKey = contextKey {
            screenPresentationConfigs[contextKey] = config
        } else {
            screenPresentationConfigs = [:]
            defaultPresentationConfig = config
        }
    }
    
    @MainActor @objc public func showScreen(_ contextKey: String) {
        NoCodes.shared.showScreen(withContextKey: contextKey)
    }
    
    @MainActor @objc public func close() {
        NoCodes.shared.close()
    }
    
    @objc public func getAvailableEvents() -> [String] {
        let availableEvents: [NoCodesEvent] = [
            .screenShown,
            .finished,
            .actionStarted,
            .actionFailed,
            .actionFinished,
            .screenFailedToLoad
        ]
        
        return availableEvents.map { $0.rawValue }
    }
}

extension NoCodesSandwich: NoCodesScreenCustomizationDelegate {
    public func presentationConfigurationForScreen(contextKey: String) -> NoCodesPresentationConfiguration {
        return screenPresentationConfigs[contextKey] ?? defaultPresentationConfig ?? .defaultConfiguration()
    }
    
    public func presentationConfigurationForScreen(id: String) -> NoCodesPresentationConfiguration {
        return screenPresentationConfigs[id] ?? defaultPresentationConfig ?? .defaultConfiguration()
    }
    
    public func viewForPopoverPresentation() -> UIView? {
        return nil
    }
}

extension NoCodesSandwich: NoCodesDelegate {
    public func controllerForNavigation() -> UIViewController? {
        return nil
    }
    
    public func noCodesHasShownScreen(id: String) {
        let payload: BridgeData = ["screenId": id]
        noCodesEventListener?.noCodesDidTrigger(event: NoCodesEvent.screenShown.rawValue, payload: payload.clearEmptyValues())
    }
    
    public func noCodesStartsExecuting(action: NoCodesAction) {
        let payload: BridgeData = action.toMap()
        noCodesEventListener?.noCodesDidTrigger(event: NoCodesEvent.actionStarted.rawValue, payload: payload.clearEmptyValues())
    }
    
    public func noCodesFailedToExecute(action: NoCodesAction, error: Error?) {
        var payload: BridgeData = action.toMap()
        payload["error"] = errorToMap(error)
        noCodesEventListener?.noCodesDidTrigger(event: NoCodesEvent.actionFailed.rawValue, payload: payload.clearEmptyValues())
    }
    
    public func noCodesFinishedExecuting(action: NoCodesAction) {
        let payload: BridgeData = action.toMap()
        noCodesEventListener?.noCodesDidTrigger(event: NoCodesEvent.actionFinished.rawValue, payload: payload.clearEmptyValues())
    }
    
    public func noCodesFinished() {
        noCodesEventListener?.noCodesDidTrigger(event: NoCodesEvent.finished.rawValue, payload: nil)
    }
    
    public func noCodesFailedToLoadScreen(error: Error?) {
        noCodesEventListener?.noCodesDidTrigger(event: NoCodesEvent.screenFailedToLoad.rawValue, payload: errorToMap(error)?.clearEmptyValues())
    }
}
#endif
